### PR TITLE
[model, ci] fix: load DeepSeek V4 Flash checkpoints

### PR DESCRIPTION
**Pull request by Luosuu**
### What does this PR do?

> Follow-up to #899. Load the released DeepSeek-V4-Flash inference checkpoint through VeOmni's streaming FSDP2 loader, including its raw inference namespace and packed FP4 expert weights.

### Checklist Before Starting

- Related work: #899
- PR title follows `[{modules}] {type}: {description}` format.

### Test

> - `make quality`
> - `pytest -q tests/models/test_checkpoint_tensor_converter.py`: 112 passed
> - `pytest -q -x tests/models/test_models_patch.py -k deepseek_v4`: 1 passed
> - Verified all 69,187 entries in `deepseek-ai/DeepSeek-V4-Flash`: 1,199 direct model tensors, 365 block-scale tensors, 66,048 expert parts, and 1,575 intentionally ignored MTP tensors; no unmatched base-model keys or shape mismatches
> - Built the full 284,332,230,231-parameter checkpoint with FSDP2 and EP=4 on four NVIDIA GB200 GPUs: all 46 shards loaded, 64 experts per rank, no remaining meta tensors, and 142.2 GB allocated per rank

### Forward Output Comparison

Compared the checkpoint's bundled official inference implementation against VeOmni on four NVIDIA GB200 GPUs using the same 10-token chat prefill for `Hello, who are you?`:

- Official path: bundled converter and inference code, MP=4, native FP4/FP8 kernels.
- VeOmni path: EP=4 + FSDP2, dequantized BF16 weights, eager attention, and fused Triton MoE.
- Both paths produced finite logits for the complete 129,280-token vocabulary.
- Both selected token `19923` (`Hello`) as the greedy next token.
- After addressing review feedback, the full EP=4 forward remained bit-for-bit identical to the pre-change VeOmni baseline (maximum absolute logit difference `0.0`); all official-comparison metrics below were unchanged.

| Metric | Result |
| --- | ---: |
| Top-5 overlap | 4/5 |
| Top-10 overlap | 9/10 |
| Mean absolute logit difference | 0.9147 |
| RMSE | 1.1315 |
| Maximum absolute logit difference | 6.5364 |
| Relative L2 error | 21.15% |
| Cosine similarity | 0.97738 |
| Pearson correlation | 0.98424 |
| KL divergence, official to VeOmni | 0.01190 |
| Jensen-Shannon divergence | 0.00254 |
| Total-variation distance | 3.40% |
| Official / VeOmni top-1 probability | 87.57% / 84.96% |

The ranking and probability-distribution agreement are strong for this prefill, while the raw logits are not bitwise equivalent because the official path executes quantized FP4/FP8 kernels and VeOmni executes dequantized BF16 weights. This short-prompt comparison does not exercise long-context compression or every routed expert.

### API and Usage Example

> N/A. Existing DeepSeek V4 model-loading APIs now accept the released Flash checkpoint format.

### Design & Code Changes

> - Reuse Transformers 5.9's registered DeepSeek V4 `WeightRenaming` rules for raw inference-to-model key conversion.
> - Unpack raw expert `I8` storage as two E2M1 FP4 values per byte before applying E8M0 block scales.
> - Restrict FP4 unpacking to raw DeepSeek expert keys, preserving conventional int8 block-scale behavior.
> - Preserve both `mtp.*` and optionally prefixed `model.mtp.*` keys for the separate MTP loading path.
> - Convert safetensor index mappings through the same namespace path and add FP8, FP4, int8, expert-fusion, and key-mapping regression coverage.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added regression tests collected by the existing model unit-test workflow

### CODE REVIEW

File: veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py [hunk 1]
```diff
@@ -76,6 +85,14 @@ def _is_quantized_weight(tensor: torch.Tensor) -> bool:
     return tensor.dtype in _QUANTIZED_WEIGHT_DTYPES
 
 
+def convert_deepseek_v4_checkpoint_key(name: str) -> str:
+    """Convert DeepSeek's inference checkpoint names to Transformers v5 names."""
+    converted_name, _ = rename_source_key(name, _DEEPSEEK_V4_WEIGHT_RENAMINGS, [])
+    if converted_name.startswith(("embed_tokens.", "hc_head.", "layers.", "norm.")):
+        converted_name = f"model.{converted_name}"
+    return converted_name
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py]: Addressed carefully in 137e142c. Optional model. is stripped before Transformers renaming, then restored only for base-model namespaces. This preserves unknown model.* keys and correctly maps model.head.weight to top-level lm_head.weight. Tests cover prefixed embedding, head, norm, hyper-head, attention, shared-expert, and unknown keys.

File: veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py [hunk 2]
```diff
@@ -88,11 +105,24 @@ def _weight_name_from_scale_name(name: str) -> str | None:
...
+        table = torch.tensor(_FP4_E2M1_VALUES, dtype=torch.float32, device=weight.device)
+        low = table[(packed & 0x0F).long()]
+        high = table[((packed >> 4) & 0x0F).long()]
+        weight = torch.stack((low, high), dim=-1).flatten(-2)
+    else:
+        weight = weight.float()
+
     if scale.numel() == 1:
-        return weight.float() * scale.float()
+        return weight * scale.float()
```
[the remark this answers is not in the document]
Comment (Luosuu) [reply veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py:159]: Addressed in 137e142c. _dequantize_scaled_weight now retrieves the per-device table from the cached helper. Validation: 108 converter tests and the DeepSeek V4 model patch test pass; the full 46-shard EP=4 forward is bit-for-bit identical to the pre-change VeOmni baseline (maximum absolute logit difference 0.0).

File: veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py [hunk 3]
```diff
@@ -76,6 +104,29 @@ def _is_quantized_weight(tensor: torch.Tensor) -> bool:
...
+
+def convert_deepseek_v4_checkpoint_key(name: str) -> str:
+    """Convert DeepSeek's inference checkpoint names to Transformers v5 names."""
+    if _is_mtp_checkpoint_key(name):
+        return name
+    has_model_prefix = name.startswith("model.")
+    source_name = name.removeprefix("model.") if has_model_prefix else name
+    converted_name, _ = rename_source_key(source_name, _DEEPSEEK_V4_WEIGHT_RENAMINGS, [])
+    if converted_name.startswith(_BASE_MODEL_KEY_PREFIXES):
+        return f"model.{converted_name}"
```
Comment (FoolPlayer) [inline veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py]: This hard-coded `model.` prefix only matches `DeepseekV4ForCausalLM`. The converter is also registered on `DeepseekV4Model`, whose state dict uses bare names such as `embed_tokens.*` and `layers.*`. Those converted keys would be skipped as unexpected, leaving the missing parameters reinitialized. Please make the prefix target-model-aware and test both model classes.

Comment (Luosuu) [reply veomni/models/transformers/deepseek_v4/checkpoint_tensor_converter.py]: Fixed in `8ffb4435`: the converter now infers the target prefix from the live model, so `DeepseekV4ForCausalLM` keeps `model.*` names while `DeepseekV4Model` emits bare base-model names. I also registered prefix-specific FQN mapping converters and added regression coverage for both layouts.

**Review by FoolPlayer — APPROVED** [review]
